File: 김은지/다단계-칫솔-판매.js
function solution(enroll, referral, seller, amount) {
  // 판매원 별 판매 이익 구하기
  const profits = new Map(enroll.map((name, idx) => [name, 0]));

  // 추천인 연결하기
  const recommender = {};
  enroll.forEach((name, idx) => {
    recommender[name] = referral[idx];
  });

  // 배분하기
  seller.forEach((name, idx) => {
    let profit = amount[idx] * 100;

    do {
      let distribution = Math.floor(profit / 10);
      let remain = profit - distribution;
      profits.set(name, profits.get(name) + remain);

      name = recommender[name];
      profit = distribution;
    } while (profit > 0 && name !== "-");
  });

  return [...profits.values()];
}

// 시간 초과
// function solution(enroll, referral, seller, amount) {
//     // 판매원 별 판매 이익 구하기
//     const profits = new Map(enroll.map((name, idx) => [name, 0]));

//     // 추천인 연결하기
//     const recommender = {};
//     enroll.forEach((name, idx) => {
//         recommender[name] = referral[idx];
//     });

//     // 배분하기
//     seller.forEach((name, idx) => {
//         let profit = amount[idx] * 100;

//         do {
//             let distribution = Math.floor(profit / 10);
//             let remain = profit - distribution;
//             profits.set(name, profits.get(name) + remain);

//             name = recommender[name];
//             profit = distribution;
//         } while (name !== '-');
//     });

//     return [...profits.values()];
// }

File: 김현경/다단계_칫솔_판매.py
def solution(enroll, referral, seller, amount):
    parent_referral = {}
    total_money = {}
    
    for i in range(len(enroll)):
        parent_referral[enroll[i]] = referral[i]
        total_money[enroll[i]] = 0 
    
    for i in range(len(seller)):
        curr_name = seller[i]
        curr_profit = amount[i] * 100
        
        while curr_name != '-' and curr_profit > 0:
            fee = curr_profit // 10
            curr_profit = curr_profit - fee
            
            total_money[curr_name] += curr_profit
            curr_name = parent_referral[curr_name]
            curr_profit = fee 
    
    return list(total_money.values())